### PR TITLE
Feature/health check

### PR DESCRIPTION
**Pull request by ivis-tsukioka**
# PULL REQUEST

## Background

* Health check of the connection to the mount storage destination was necessary

## Main Points of Modification

*　Added file storage connection confirmation function
　http://localhost/healthcheck is OK
image
　http://localhost/healthcheck is NG
image

* 　A value in the app.ini file (dg.HEALTH_FILE_PATH and dg.HEALTH_FILE_NAME) activates the health check, and the absence of either value disables it.

*  At startup of Gin, if dg.HEALTH_FILE_PATH and dg.HEALTH_FILE_NAME values are present, a file for health check is created in a specific path. If there is no value, the file is skipped. It is also skipped if the file already exists.

if in app.ini
```
[dg]
HEALTH_FILE_PATH = /data/gogs/mnt/health
HEALTH_FILE_NAME = health.txt
```
→create /data/gogs/mnt/health/health.txt

## Test

* いくつかのケースにおいて上記の機能が正常に動くことを確認した。
  1. Initial startup, with value
  2. Reboot, with value
  3. Initial startup, no value
  4. Reboot, no value

## Viewpoint for Review

* None

## Supplementary Information

* To enable file system health checks, include the following in app.ini
```
[dg]
HEALTH_FILE_PATH = mount path
HEALTH_FILE_NAME = file name
```

## ToDo

* None

### CODE REVIEW

**Review by ivis-akaike — COMMENTED** [review]
誤記ですかね。
誤：exits
正：exist

File: internal/cmd/web.go [hunk 1]
```diff
@@ -36,6 +36,7 @@ import (
 	"github.com/NII-DG/gogs/internal/context"
```
Comment (ivis-tsukioka) [inline internal/cmd/web.go:36]: コメントが見当たりません。

**Review by ivis-tsukioka — COMMENTED** [review]
test

File: internal/healthcheck/filesystem.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+func CheckFileSystem() error {
+	if len(conf.DG.HealthFilePath) <= 0 || len(conf.DG.HealthFileName) <= 0 {
+		log.Trace("health check to file system is skip")
+		return nil
+	}
+	filepath := filepath.Join(conf.DG.HealthFilePath, conf.DG.HealthFileName)
+	if utils.ExistData(filepath) {
+		return nil
+	} else {
+		return fmt.Errorf("Not exits %s for health check", filepath)
```
[the remark this answers is not in the document]
Comment (ivis-tsukioka) [reply internal/healthcheck/filesystem.go]: 修正しました。

File: internal/route/install.go [hunk 3]
```diff
@@ -415,3 +422,24 @@ func InstallPost(c *context.Context, f form.Install) {
...
+	dirpath := conf.DG.HealthFilePath
+	if !utils.ExistData(dirpath) {
+		if err := os.Mkdir(dirpath, os.ModePerm); err != nil {
+			return fmt.Errorf("failure to create directory. dir : %s, err: %v", dirpath, err)
+		}
+	}
+	filepath := filepath.Join(dirpath, conf.DG.HealthFileName)
+	file, err := os.OpenFile(filepath, os.O_WRONLY|os.O_CREATE, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failure to create File. filepaht : %s, err: %v", filepath, err)
```
[the remark this answers is not in the document]
Comment (ivis-tsukioka) [reply internal/route/install.go]: 修正しました。

**Pull request merged**